external-crates/move/crates/move-binary-format/src/normalized.rs

```diff
@@ -65,81 +74,133 @@ pub enum Type {
     U256,
 }
 
-/// Normalized version of a `FieldDefinition`. The `name` is included even though it is
-/// metadata that it is ignored by the VM. The reason: names are important to clients. We would
-/// want a change from `Account { bal: u64, seq: u64 }` to `Account { seq: u64, bal: u64 }` to be
-/// marked as incompatible. Not safe to compare without an enclosing `Struct`.
-#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
-pub struct Field {
-    pub name: Identifier,
-    pub type_: Type,
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Datatype<S> {
+    pub address: AccountAddress,
+    pub module: S,
+    pub name: S,
+    pub type_arguments: Vec<Type<S>>,
+}
+
+pub type Signature<S> = Rc<Vec<Rc<Type<S>>>>;
+
+#[cfg_attr(test, derive(Clone))]
+#[derive(Debug)]
+struct Tables<S> {
+    empty_signature: Signature<S>,
+    signatures: Vec<Signature<S>>,
+    constants: Vec<Rc<Constant<S>>>,
+    struct_defs: Vec<Rc<Struct<S>>>,
+    function_defs: Vec<Rc<Function<S>>>,
+    enum_defs: Vec<Rc<Enum<S>>>,
+}
+
+/// Normalized version of a `CompiledModule`: its address, name, struct declarations, and public
+/// function declarations.
+#[cfg_attr(test, derive(Clone))]
+#[derive(Debug)]
+pub struct Module<S: Ord> {
+    #[allow(unused)]
+    tables: Tables<S>,
+    pub id: ModuleId<S>,
+    pub file_format_version: u32,
+    pub dependencies: Vec<ModuleId<S>>,
+    pub friends: Vec<ModuleId<S>>,
+    pub structs: BTreeMap<S, Rc<Struct<S>>>,
+    pub enums: BTreeMap<S, Rc<Enum<S>>>,
+    pub functions: BTreeMap<S, Rc<Function<S>>>,
+    pub constants: Vec<Rc<Constant<S>>>,
 }
 
 /// Normalized version of a `Constant`.
-#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
-pub struct Constant {
-    pub type_: Type,
+#[cfg_attr(test, derive(Clone))]
```

Comment:
I removed `Clone` from all these types (except for tests) to prevent issues of accidentally cloning the type instead of the outer `Rc`